keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,469 @@
+- [Proposal](#proposal)
+  - [Handle the Return of Shutdown Node](#handle-the-return-of-shutdown-node)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Unit tests](#unit-tests)
+  - [E2E tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha](#alpha)
+    - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+    - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+  - [rely on the taint as a lock mechanism (taint and untaint by the same actor)](#rely-on-the-taint-as-a-lock-mechanism-taint-and-untaint-by-the-same-actor)
+  - [Node fencing](#node-fencing)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
+approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+In case of a node being shutdown (i.e. not running) the control plane doesn't make the right
+assumptions to enable stateful workload to fail-over. This KEP introduces a flow to do so.
+
+## Motivation
```

Comment:
> The Graceful Node Shutdown KEP talks about adding ShutdownGracePeriod for Pods. It didn't talk about how to handle VolumeAttachment.

I would like to double click on this. Reading the graceful node shutdown blog, it looks like what the feature does is delay the shutdown of a node to give a longer grace period for pods to gracefully terminate. If that's the case, I would expect that grace period to also allow time to unmount and detach volumes from the node.

If that's not working, then I agree with @mkimuram that this may be a bug in the feature, and we should investigate how to fix it. 

This proposal seems to be handling the scenario where perhaps the grace period has passed and the node has ended up shutting down before we are able to finish tearing down the volumes. But I think in terms of priority, we should focus on getting the graceful sequence to work first.

Reply:
It is not that we don't have enough time to detach the volumes. The detach won't happen unless if we make the code changes suggested in this KEP. I've tested this by shutting down one node and observing the pods of a StatefulSet won't move to another node, even after enabling the Graceful Node Shutdown feature (and also set those two periods to > 0).

@bobbypage Can you chime in?

Reply:
The current behavior of graceful node shutdown feature is upon shutdown kubelet terminates the pod and puts it into the failed phase. There's wasn't any special logic applied for volume detachment, but it's great to see this KEP consider how we can leverage graceful node shutdown for volume detaching upon shutdown. 

I chatted with @msau42 offline, it seemed like there's two steps in the volume detaching process:

1) kubelet unmounting the volume
2) signaling to kube-controller-manager that the volume is safe to detach

For #1, it seems like volume detachment is managed by kubelet and it will detach if the pod is deleted and containers have finished cleaning up [1]. Since the graceful node shutdown terminates the pod, I would expect eventually that kubelet volume manager will see the pod is terminated and that the detach would happen. We should confirm this is in fact the case.

For #2, as per my chat with @msau42 it depends on the pod having a deletion timestamp set which graceful node shutdown does not currently do. 

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L244

Reply:
@bobbypage can you also clarify when you say "upon shutdown" do you mean "when the node gets a shutdown signal but in the grace period before the system actually goes down"

Reply:
> @bobbypage can you also clarify when you say "upon shutdown" do you mean "when the node gets a shutdown signal but in the grace period before the system actually goes down"

Yes. 

The logic is that kubelet is watching system shutdown events from systemd. When kubelet is notified of the shutdown event, it uses a systemd inhibitor lock to request additional time before the system gets shutdown. This duration of additional grace period is configured in KubeletConfig as `ShutdownGracePeriod`. Kubelet uses this additional grace period to gracefully terminate pods.

Reply:
> Since the graceful node shutdown terminates the pod, I would expect eventually that kubelet volume manager will see the pod is terminated and that the detach would happen. We should confirm this is in fact the case.

Current behavior seems just a bug in k8s 1.22 and it worked well with k8s 1.21. 
Please see my comments [here](https://github.com/kubernetes/kubernetes/issues/104531#issuecomment-912717050) and [here](https://github.com/kubernetes/kubernetes/issues/104531#issuecomment-912822663).

Reply:
> For #2, as per my chat with @msau42 it depends on the pod having a deletion timestamp set which graceful node shutdown does not currently do. During graceful node shutdown, kubelet terminates the pod and puts the pod into failed phase, but does not delete the pod or set deletion timestamp.

How "eventually" ?  I've heard from many people that this long delay is a problem.

Reply:
For the graceful node shutdown case, it is expected that volume teardown happens with the same time expectations as a normal pod deletion. It does not appear to be happening in this case, so we're focusing right now on understanding the graceful node shutdown design in more detail and its impact on kubelet volume manager and AD controller.

Reply:
There are two main issues with the Graceful Node Shutdown feature:
1. Whether Kubelet can detect a “Graceful node shutdown” depends on which shutdown command is used. In my testing on Ubuntu, “systemctl poweroff” is the only command that is detected by Kubelet as a “Graceful node shutdown”. Other commands such as “shutdown -h now” or shutting down a VM from vSphere UI are not detected by Kubelet, so they all fall into “Non-graceful node shutdown” cases.
2. The bug in k8s 1.22 mentioned above. In k8s 1.21, Pod status is changed to “Failed” when Kubelet detects the node shutdown so volumes are detached properly and pods are moved to a remaining running node. In k8s 1.22, the Pod status is not changed to “Failed” and stays as “Running” when Kubelet detects the node shutdown, so volumes stay as attached and pods are not moved to a remaining running node.